File: packages/ESMFormat.jl/src/reactions.jl
"""
Reaction system to ODE conversion functionality.

This module implements the conversion of ReactionSystem objects to ODE-based Models
using standard mass action kinetics as specified in Libraries Spec Section 7.4.
"""

"""
    derive_odes(rxn_sys::ReactionSystem) -> Model

Generate an ODE model from a reaction system using standard mass action kinetics.

Creates state variables for each species, parameters for rate constants, and equations
with d[species]/dt = net_stoichiometry * rate_law. Handles source reactions (null
substrates), sink reactions (null products), and constraint equations.

# Arguments
- `rxn_sys::ReactionSystem`: Input reaction system with species, reactions, and parameters

# Returns
- `Model`: ODE model with variables, equations, and optional events/subsystems

# Algorithm (Libraries Spec Section 7.4)
1. Create state variables for all species in the reaction system
2. Create parameter variables for all parameters (rate constants)
3. For each reaction, compute the mass action rate expression
4. For each species, compute the net rate of change from all reactions
5. Create differential equations d[species]/dt = net_rate

# Examples
```julia
# Simple A + B -> C reaction
species = [Species("A"), Species("B"), Species("C")]
reactions = [Reaction(Dict("A"=>1, "B"=>1), Dict("C"=>1), VarExpr("k1"))]
parameters = [Parameter("k1", 0.1)]
rxn_sys = ReactionSystem(species, reactions, parameters=parameters)

model = derive_odes(rxn_sys)
# Returns Model with:
# - State variables: A, B, C
# - Parameter: k1
# - Equations: d[A]/dt = -k1*A*B, d[B]/dt = -k1*A*B, d[C]/dt = k1*A*B
```
"""
function derive_odes(rxn_sys::ReactionSystem)::Model
    # Collect all species names
    species_names = [species.name for species in rxn_sys.species]

    # Create state variables for all species
    variables = Dict{String,ModelVariable}()
    for species in rxn_sys.species
        variables[species.name] = ModelVariable(
            StateVariable,
            default=1.0,  # Default initial concentration
            description="Concentration of species $(species.name)",
            units="mol/L"
        )
    end

    # Add parameter variables for all parameters
    for param in rxn_sys.parameters
        variables[param.name] = ModelVariable(
            ParameterVariable,
            default=param.default,
            description=param.description,
            units=param.units
        )
    end

    # Compute stoichiometric matrix
    S = stoichiometric_matrix(rxn_sys)

    # Create differential equations for each species
    equations = Equation[]

    for (i, species_name) in enumerate(species_names)
        # Create left-hand side: d[species]/dt
        lhs = OpExpr("D", ESMFormat.Expr[VarExpr(species_name)], wrt="t")

        # Create right-hand side: sum of (stoichiometry * reaction_rate) for all reactions
        rate_terms = ESMFormat.Expr[]

        for (j, reaction) in enumerate(rxn_sys.reactions)
            stoich = S[i, j]
            if stoich != 0
                # Get mass action rate for this reaction
                rate_expr = mass_action_rate(reaction, rxn_sys.species)

                if stoich == 1
                    push!(rate_terms, rate_expr)
                elseif stoich == -1
                    push!(rate_terms, OpExpr("-", ESMFormat.Expr[rate_expr]))
                else
                    push!(rate_terms, OpExpr("*", ESMFormat.Expr[NumExpr(Float64(stoich)), rate_expr]))
                end
            end
        end

        # Combine all rate terms
        rhs = if isempty(rate_terms)
            NumExpr(0.0)
        elseif length(rate_terms) == 1
            rate_terms[1]
        else
            OpExpr("+", rate_terms)
        end

        push!(equations, Equation(lhs, rhs))
    end

    # Handle subsystems recursively
    subsystems = Dict{String,Model}()
    for (name, subsys) in rxn_sys.subsystems
        subsystems[name] = derive_odes(subsys)
    end

    return Model(variables, equations, subsystems=subsystems)
end

"""
    stoichiometric_matrix(rxn_sys::ReactionSystem) -> Matrix{Int}

Compute the net stoichiometric matrix (species × reactions).

The stoichiometric matrix S has dimensions (n_species × n_reactions) where:
- S[i,j] = net stoichiometry of species i in reaction j
- Net stoichiometry = products - reactants
- Positive values indicate production, negative values indicate consumption

# Arguments
- `rxn_sys::ReactionSystem`: Input reaction system

# Returns
- `Matrix{Int}`: Stoichiometric matrix with dimensions (n_species, n_reactions)

# Examples
```julia
# For reaction A + B -> 2C:
# S = [-1  # A consumed
#      -1  # B consumed
#       2] # C produced (2 molecules)
```
"""
function stoichiometric_matrix(rxn_sys::ReactionSystem)::Matrix{Int}
    n_species = length(rxn_sys.species)
    n_reactions = length(rxn_sys.reactions)

    # Create species name to index mapping
    species_idx = Dict(species.name => i for (i, species) in enumerate(rxn_sys.species))

    # Initialize stoichiometric matrix
    S = zeros(Int, n_species, n_reactions)

    for (j, reaction) in enumerate(rxn_sys.reactions)
        # Handle substrates (negative stoichiometry)
        if reaction.substrates !== nothing
            for entry in reaction.substrates
                if haskey(species_idx, entry.species)
                    i = species_idx[entry.species]
                    S[i, j] -= entry.stoichiometry
                end
            end
        end

        # Handle products (positive stoichiometry)
        if reaction.products !== nothing
            for entry in reaction.products
                if haskey(species_idx, entry.species)
                    i = species_idx[entry.species]
                    S[i, j] += entry.stoichiometry
                end
            end
        end
    end

    return S
end

"""
    mass_action_rate(reaction::Reaction, species::Vector{Species}) -> Expr

Compute the mass action rate expression for a reaction.

Uses standard mass action kinetics: rate = k * ∏[reactants]^stoichiometry
Handles special cases:
- Source reactions (no reactants): rate = k
- Sink reactions (no products): rate = k * ∏[reactants]^stoichiometry
- Reversible reactions: rate = k_forward * ∏[reactants]^stoich - k_reverse * ∏[products]^stoich

# Arguments
- `reaction::Reaction`: The reaction to compute the rate for
- `species::Vector{Species}`: All species in the system (for validation)

# Returns
- `Expr`: Mass action rate expression

# Examples
```julia
# For reaction A + B -> C with rate k:
# Returns: k * A * B

# For source reaction -> A with rate k:
# Returns: k

# For reversible reaction A ⇌ B with rate k:
# Returns: k * A (if reaction.reversible = false)
```
"""
function mass_action_rate(reaction::Reaction, species::Vector{Species})::ESMFormat.Expr
    # Start with the rate constant expression
    rate_expr = reaction.rate

    # Handle source reactions (no substrates)
    if reaction.substrates === nothing || isempty(reaction.substrates)
        return rate_expr
    end

    # Build mass action terms for substrates
    mass_action_terms = ESMFormat.Expr[rate_expr]

    for entry in reaction.substrates
        species_expr = VarExpr(entry.species)
        if entry.stoichiometry == 1
            push!(mass_action_terms, species_expr)
        else
            # For stoichiometry > 1, use power operator
            push!(mass_action_terms, OpExpr("^", ESMFormat.Expr[species_expr, NumExpr(Float64(entry.stoichiometry))]))
        end
    end

    # Combine all terms with multiplication
    if length(mass_action_terms) == 1
        return mass_action_terms[1]
    else
        return OpExpr("*", mass_action_terms)
    end
end

File: packages/esm_format/src/esm_format/reactions.py
"""
Reaction system analysis and ODE generation functions.

This module provides functions to:
1. Generate ODE systems from reaction systems using mass action kinetics
2. Compute stoichiometric matrices for reaction networks
3. Handle substrate and product matrices separately
"""

import numpy as np
import sympy as sp
from typing import Dict, List, Optional, Set, Tuple

from .esm_types import ReactionSystem, Reaction, Species, Model, ModelVariable, Equation, Expr, ExprNode
from .expression import to_sympy, from_sympy


def derive_odes(system: ReactionSystem) -> Model:
    """
    Derive ODEs from a reaction system using mass action kinetics.

    Generates a system of ordinary differential equations from a reaction network
    based on the stoichiometry and rate laws. Handles source reactions (null
    substrates) and sink reactions (null products).

    Args:
        system: ReactionSystem containing species, reactions, and parameters

    Returns:
        Model: Mathematical model with ODE equations for each species

    Raises:
        ValueError: If system is empty or contains invalid reactions
    """
    if not system.species:
        raise ValueError("ReactionSystem must contain at least one species")

    if not system.reactions:
        raise ValueError("ReactionSystem must contain at least one reaction")

    # Create species concentration variables
    species_names = [species.name for species in system.species]
    variables = {}

    # Add state variables for species concentrations
    for species in system.species:
        variables[species.name] = ModelVariable(
            type='state',
            units=species.units,
            description=f"Concentration of {species.name}",
        )

    # Add parameter variables
    for param in system.parameters:
        variables[param.name] = ModelVariable(
            type='parameter',
            units=param.units,
            description=param.description,
            default=param.value if isinstance(param.value, (int, float)) else None,
            expression=param.value if not isinstance(param.value, (int, float)) else None,
        )

    # Generate ODE equations using mass action kinetics
    equations = []

    # Initialize rate of change for each species
    species_rates = {name: 0 for name in species_names}

    for reaction in system.reactions:
        # Build rate expression using mass action kinetics
        # Rate = k * [reactant1]^coeff1 * [reactant2]^coeff2 * ...

        if reaction.rate_constant is None:
            raise ValueError(f"Reaction {reaction.name} must have a rate constant")

        rate_expr = reaction.rate_constant

        # Add reactant terms (mass action kinetics)
        for reactant, coeff in reaction.reactants.items():
            if reactant not in species_names:
                raise ValueError(f"Reactant {reactant} not found in species list")

            # For mass action: rate *= [reactant]^coeff
            if coeff == 1:
                # Simple multiplication
                rate_expr = _multiply_expressions(rate_expr, reactant)
            else:
                # Power term
                power_expr = _power_expression(reactant, coeff)
                rate_expr = _multiply_expressions(rate_expr, power_expr)

        # Apply stoichiometric coefficients to species rates
        # For each species, d[species]/dt += (product_coeff - reactant_coeff) * rate

        for species_name in species_names:
            net_stoich_coeff = 0

            # Subtract reactant coefficient (consumed)
            if species_name in reaction.reactants:
                net_stoich_coeff -= reaction.reactants[species_name]

            # Add product coefficient (produced)
            if species_name in reaction.products:
                net_stoich_coeff += reaction.products[species_name]

            if net_stoich_coeff != 0:
                # Add this reaction's contribution to species rate
                contribution = _multiply_expressions(net_stoich_coeff, rate_expr)
                species_rates[species_name] = _add_expressions(species_rates[species_name], contribution)

    # Create differential equations
    for species_name in species_names:
        if species_rates[species_name] != 0:  # Only create equations for species with non-zero rates
            # d[species]/dt = rate_expression
            lhs = ExprNode(op="D", args=[species_name], wrt="t")
            equations.append(Equation(lhs=lhs, rhs=species_rates[species_name]))

    # Handle constraint equations (algebraic equations)
    # These would be derived from conservation laws or equilibrium assumptions
    # For now, we don't add any constraint equations - they would be handled
    # separately based on the specific system requirements

    return Model(
        name=f"{system.name}_odes",
        variables=variables,
        equations=equations,
        metadata={
            "derived_from": system.name,
            "generation_method": "mass_action_kinetics"
        }
    )


def stoichiometric_matrix(system: ReactionSystem) -> np.ndarray:
    """
    Compute the net stoichiometric matrix for a reaction system.

    The stoichiometric matrix S is defined such that S[i,j] gives the net
    stoichiometric coefficient of species i in reaction j. Negative values
    indicate reactants (consumed), positive values indicate products (formed).

    Args:
        system: ReactionSystem containing species and reactions

    Returns:
        np.ndarray: Stoichiometric matrix with shape (n_species, n_reactions)
                   where S[i,j] = net stoichiometric coefficient of species i in reaction j

    Raises:
        ValueError: If system is empty
    """
    if not system.species or not system.reactions:
        return np.array([])

    species_names = [s.name for s in system.species]
    n_species = len(species_names)
    n_reactions = len(system.reactions)

    matrix = np.zeros((n_species, n_reactions))

    for j, reaction in enumerate(system.reactions):
        for i, species_name in enumerate(species_names):
            net_coeff = 0

            # Subtract reactant coefficient (negative contribution)
            if species_name in reaction.reactants:
                net_coeff -= reaction.reactants[species_name]

            # Add product coefficient (positive contribution)
            if species_name in reaction.products:
                net_coeff += reaction.products[species_name]

            matrix[i, j] = net_coeff

    return matrix


def substrate_matrix(system: ReactionSystem) -> np.ndarray:
    """
    Compute the substrate (reactant) stoichiometric matrix.

    The substrate matrix gives only the reactant stoichiometric coefficients.
    All entries are non-negative, with S[i,j] giving the coefficient of
    species i as a reactant in reaction j.

    Args:
        system: ReactionSystem containing species and reactions

    Returns:
        np.ndarray: Substrate matrix with shape (n_species, n_reactions)
                   where S[i,j] = reactant coefficient of species i in reaction j
    """
    if not system.species or not system.reactions:
        return np.array([])

    species_names = [s.name for s in system.species]
    n_species = len(species_names)
    n_reactions = len(system.reactions)

    matrix = np.zeros((n_species, n_reactions))

    for j, reaction in enumerate(system.reactions):
        for i, species_name in enumerate(species_names):
            if species_name in reaction.reactants:
                matrix[i, j] = reaction.reactants[species_name]

    return matrix


def product_matrix(system: ReactionSystem) -> np.ndarray:
    """
    Compute the product stoichiometric matrix.

    The product matrix gives only the product stoichiometric coefficients.
    All entries are non-negative, with P[i,j] giving the coefficient of
    species i as a product in reaction j.

    Args:
        system: ReactionSystem containing species and reactions

    Returns:
        np.ndarray: Product matrix with shape (n_species, n_reactions)
                   where P[i,j] = product coefficient of species i in reaction j
    """
    if not system.species or not system.reactions:
        return np.array([])

    species_names = [s.name for s in system.species]
    n_species = len(species_names)
    n_reactions = len(system.reactions)

    matrix = np.zeros((n_species, n_reactions))

    for j, reaction in enumerate(system.reactions):
        for i, species_name in enumerate(species_names):
            if species_name in reaction.products:
                matrix[i, j] = reaction.products[species_name]

    return matrix


# Helper functions for expression manipulation

def _multiply_expressions(expr1: Expr, expr2: Expr) -> Expr:
    """Multiply two expressions."""
    if expr1 == 0:
        return 0
    if expr2 == 0:
        return 0
    if expr1 == 1:
        return expr2
    if expr2 == 1:
        return expr1

    return ExprNode(op="*", args=[expr1, expr2])


def _add_expressions(expr1: Expr, expr2: Expr) -> Expr:
    """Add two expressions."""
    if expr1 == 0:
        return expr2
    if expr2 == 0:
        return expr1

    return ExprNode(op="+", args=[expr1, expr2])


def _power_expression(base: Expr, exponent: float) -> Expr:
    """Create a power expression."""
    if exponent == 1:
        return base
    if exponent == 0:
        return 1

    return ExprNode(op="^", args=[base, exponent])